File: rust/src/leetcode/Q695_max_area_of_island.rs
/*
 * @lc app=leetcode id=695 lang=rust
 *
 * [695] Max Area of Island
 */

// @lc code=start
impl Solution {
    fn dfs(row: i32, col: i32, grid: &mut Vec<Vec<i32>>) -> i32 {
        if row < 0
            || col < 0
            || row >= grid.len() as i32
            || col >= grid[0].len() as i32
            || grid[row as usize][col as usize] == 0
        {
            return 0;
        }

        let mut size: i32 = 1;
        grid[row as usize][col as usize] = 0;

        size += Self::dfs(row - 1, col, grid);
        size += Self::dfs(row + 1, col, grid);
        size += Self::dfs(row, col - 1, grid);
        size += Self::dfs(row, col + 1, grid);

        size
    }

    pub fn max_area_of_island(grid: Vec<Vec<i32>>) -> i32 {
        let mut result: i32 = 0;
        let mut grid: Vec<Vec<i32>> = grid;

        for i in 0..grid.len() {
            for j in 0..grid[0].len() {
                let size: i32 = Self::dfs(i as i32, j as i32, &mut grid);

                if size > result {
                    result = size;
                }
            }
        }

        result
    }
}
// @lc code=end

File: python/Q695_max_area_of_island.py
#
# @lc app=leetcode id=695 lang=python3
#
# [695] Max Area of Island
#


# @lc code=start
class Solution:
    def dfs(self, row: int, col: int) -> int:
        if (
            row < 0
            or col < 0
            or row >= self.ROWS
            or col >= self.COLS
            or self.grid[row][col] == 0
        ):
            return 0

        area: int = 1

        self.grid[row][col] = 0
        for x, y in self.directions:
            area += self.dfs(row + x, col + y)

        return area

    def maxAreaOfIsland(self, grid: list[list[int]]) -> int:
        self.ROWS: int = len(grid)
        self.COLS: int = len(grid[0])
        self.grid: list[list[int]] = grid

        self.max_area: int = 0

        self.directions: list[tuple[int, int]] = [(1, 0), (-1, 0), (0, 1), (0, -1)]

        for x in range(self.ROWS):
            for y in range(self.COLS):
                if (new_area := self.dfs(x, y)) > self.max_area:
                    self.max_area = new_area

        return self.max_area


# @lc code=end
